themes/default/content/blog/safe-lambda-secrets/index.md

```diff
@@ -0,0 +1,545 @@
+
+```typescript
+const secretManagerPolicyDoc = aws.iam.getPolicyDocumentOutput({
+    statements: [
+        {
+            effect: "Allow",
+            actions: [
+                "secretsmanager:GetSecretValue",
+            ],
+            resources: secretArray.map(x => pulumi.interpolate`${x.arn}`)
+        }
+    ]
+});
+
+const secretManagerPolicy = new aws.iam.Policy("secretsPolicy", {
+    policy: secretManagerPolicyDoc.apply(doc => doc.json)
+});
+```
+
+Finally we're going to attach the above policy and the AWS Managed `AWSLambdaBasicExecutionRole` policy to the IAM role we created:
+
+```typescript
+new aws.iam.RolePolicyAttachment("rpa-basic", {
+    role: role,
+    policyArn: aws.iam.ManagedPolicy.AWSLambdaBasicExecutionRole
+});
+
+new aws.iam.RolePolicyAttachment("rpa-secrets", {
+    role: role,
+    policyArn: secretManagerPolicy.arn
+});
+```
+
+The `aws.iam.ManagedPolicy.AWSLambdaBasicExecutionRole` is one of our enums that points at the relevant ARN for this policy.
+
+### Finally, the Function
+
+There are two parts to this: the `Lambda Function` resource and the code we're going to deploy as part of the function.
+
+#### The Resource
+
+The function resource itself is pretty simple. The interesting bit is how we take the array of the ARNs we created earlier, combine it with the array of environment variable values we have in the config and assign them to the `environment` input:
+
+```typescript
+let lambdaEnvVars: pulumi.Input<{
+    [key: string]: pulumi.Input<string>;
+}> = {};
+
+for (let key in secretsArnArray) {
+    lambdaEnvVars[key.toLocaleUpperCase()] = secretsArnArray[key];
+}
+
+for (let key in lambdaconfig.envvars) {
+    lambdaEnvVars[key.toLocaleUpperCase()] = lambdaconfig.envvars[key];
+}
+```
+
+The reason that we're using the `.toLocaleUpperCase()` function is because as a rule, the usual way of defining environment variable names is using upper case letters.
+
+Once we've done this we can put together the function resource:
+
+```typescript
+const lambda = new aws.lambda.Function("lambdaWithSecrets", {
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    role: role.arn,
+    handler: "index.handler",
+    runtime: aws.lambda.Runtime.NodeJS16dX,
+    environment: {
+        variables: lambdaEnvVars
+    },
+    timeout: 15
+})
+
+export const lambdaName = lambda.name;
+```
+
+There's nothing complicated about the function. We're using the role from earlier, setting the environment variables to be the object we created just above and pointed the `code` input at a local folder called `app` which we'll look at next. The lambda name being exported at the end is so that we can use the Pulumi CLI to get this value and invoke the function using the AWS CLI.
```

Comment:
```suggestion
There's nothing complicated about the function. We're using the role from earlier, setting the environment variables to be the object we created just above and pointed the `code` input at a local folder called `app`, which we'll look at next. The lambda name being exported at the end is so that we can use the Pulumi CLI to get this value and invoke the function using the AWS CLI.
```